pyfr/plugins/fwh.py

```diff
@@ -0,0 +1,197 @@
+            self.qinf['c'] = (gamma * self.qinf['p'] / self.qinf['rho'])**0.5
+
+        self.qinf['M'] = [self.qinf[k]/self.qinf['c'] for k in 'uvw'[:ndims]]
+        self.Minf = np.linalg.norm(self.qinf['M'])
+
+        ele_map = intg.system.ele_map
+        self.emap = {k: i for i, k in enumerate(ele_map)}
+        self._init_surf(intg, ele_map)
+
+    def _init_surf(self, intg, ele_map):
+        ndims = self.ndims
+
+        self.prepare_surface_data(intg)
+
+        self.proj = {}
+        for etype, v in self._surf_region_data.items():
+            eles = ele_map[etype]
+
+            pnorm = eles._pnorm_fpts
+            rcpdjacs = eles.rcpdjac_at_np('fpts')
+
+            for fidx in range(len(eles.basis.faces)):
+                eidx = [i for i, x in v if x == fidx]
+                facefpts = eles.basis.facefpts[fidx]
+                fslice = np.ix_(facefpts, eidx)
+
+                m0 = eles.basis.m0[facefpts]
+                a = eles.basis.fpts_wts[facefpts][:, None] / rcpdjacs[fslice]
+                a = a.flat
+                n_mag = np.linalg.norm(pnorm[fslice], axis=-1, keepdims=True)
+                n = (pnorm[fslice] / n_mag).transpose(2, 0, 1).reshape((ndims,
+                                                                        -1))
+                dist = self._distances(eles, fslice)
+
+                self.proj[etype, fidx] = (eidx, m0, n, a, dist)
+
+    def _distances(self, eles, fslice):
+        spts = eles.ploc_at_np('fpts').transpose(0, 2, 1)[fslice]
+        s = spts.reshape(-1, self.ndims)
+        o = self.obsv_pts
+
+        r_o = o[None, :] - s[:, None]
+        d = np.linalg.norm(r_o, axis=-1)
+        x = r_o / d[:, :, None]
+
+        m_r = self.qinf['M']*x
+        m_r_dot = np.sum(m_r, axis=-1)
+
+        inv_gamma = np.sqrt(1 - self.Minf**2)
+
+        gamma = 1 / inv_gamma
+        s_fact = (inv_gamma*inv_gamma + m_r_dot*m_r_dot)**0.5
+        r_star_vec = r_o*s_fact[..., None]
+        r_vec = gamma**2*(r_star_vec - r_o*m_r_dot[..., None])
+
+        r_star = np.linalg.norm(r_star_vec, axis=-1)
+        r = np.linalg.norm(r_vec, axis=-1)
+
+        r_star_vec_n = r_star_vec / r_star[..., None]
+        r_vec_n = r_vec / r[..., None]
+
+        return r_vec_n, 1 / r_star, r_star_vec_n
+
+    def fwh_solve(self, intg):
+        o_vals = np.zeros(self.nobvs)
+        ci = 1 / self.qinf['c']
+
+        soln, soln_t = {}, {}
+        # Querry solutions all at once to prevent MPI deadlock
+        for (etype, fidx), p in self.proj.items():
+            eidx, m0, n, a, (R, r_star_inv, R_star) = p
+            s = intg.soln[self.emap[etype]][..., eidx]
+            soln[etype, fidx] = m0 @ s.transpose(1, 0, 2)
+
+            s_t = intg.dt_soln[self.emap[etype]][..., eidx]
+            soln_t[etype, fidx] = m0 @ s_t.transpose(1, 0, 2)
+
+        # Accumulate FWH contribution from each surface part
+        for (etype, fidx), p in self.proj.items():
+            eidx, m0, n, a, (R, r_star_inv, R_star) = p
```

Comment:
Not sure if proj is the best variable name.  Also not sure if we need the second level of packing.

Reply:
It might be worth stuffing these into either a nametuple or a dataclass.